File: ui-spikes/src/main/scala/ui/todo/components/Link.scala
package ui.todo.components

import typings.reactLib.dsl._
import typings.reactLib.reactMod.{ButtonHTMLAttributes, CSSProperties, FC, HTMLAttributes}
import ui.todo.context.Context
import ui.todo.models.VisibilityFilter

import scala.language.implicitConversions
import scala.scalajs.js

object Link {

  class Props(val filter: VisibilityFilter) extends js.Object
  object Props {
    def apply(filter: VisibilityFilter): Props = new Props(filter)
  }

  val Component: FC[Props] = define.fc[Props] { props =>
    println(s"**** rendering Link filter=${props.filter}")

    val (filter, setFilter) = Context.VisibilityFilter.use()

    button.props(
      ButtonHTMLAttributes(
        HTMLAttributes(
          onClick = _ => setFilter(props.filter),
          `aria-disabled` = props.filter == filter,
          style = new CSSProperties {
            marginLeft = "4px"
          }
        )
      ),
      props.children.getOrElse(null)
    )
  }
}

File: spikes/todo/src/components/Link.js
import React, {useContext} from 'react'
import PropTypes from 'prop-types'
import VisibilityFilterContext from "../context/VisibilityFilterContext";


const Link = (props) => {
    const context = useContext(VisibilityFilterContext)

    return (
        <button
            onClick={() => context.setVisibilityFilter(props.filter)}
            disabled={props.filter === context.visibilityFilter}
            style={{
                marginLeft: '4px'
            }}
        >
            {props.children}
        </button>
    )

}

Link.propTypes = {
    filter: PropTypes.string.isRequired,
    children: PropTypes.node.isRequired,
}

export default Link
